samples/nrf_desktop/src/modules/config.c

```diff
@@ -0,0 +1,293 @@
+			LOG_INF("Store %d in %s", options.time_run,
+				MODULE_NAME "/time_run");
+			err = settings_save_one(MODULE_NAME "/time_run",
+						&options.time_run,
+						sizeof(options.time_run));
+		}
+
+		break;
+
+	case CONFIG_EVENT_ID_MOUSE_DOWNSHIFT_REST1:
+		if (options.time_rest1 != sys_get_le32(data)) {
+			options.time_rest1 = sys_get_le32(data);
+			LOG_INF("Store %d in %s", options.time_rest1,
+				MODULE_NAME "/time_rest1");
+			err = settings_save_one(MODULE_NAME "/time_rest1",
+						&options.time_rest1,
+						sizeof(options.time_rest1));
+		}
+
+		break;
+
+	case CONFIG_EVENT_ID_MOUSE_DOWNSHIFT_REST2:
+		if (options.time_rest2 != sys_get_le32(data)) {
+			options.time_rest2 = sys_get_le32(data);
+			LOG_INF("Store %d in %s", options.time_rest2,
+				MODULE_NAME "/time_rest2");
+			err = settings_save_one(MODULE_NAME "/time_rest2",
+						&options.time_rest2,
+						sizeof(options.time_rest2));
+		}
+
+		break;
+
+	default:
+		LOG_WRN("Config ID not supported");
+
+		break;
+	}
+
+	return err;
+}
+
+static int enable_settings(void)
+{
+	int err = 0;
+
+	if (IS_ENABLED(CONFIG_SETTINGS)) {
+		if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+			static struct settings_handler sh = {
+				.name = MODULE_NAME,
+				.h_set = settings_set,
+				.h_commit = commit,
+			};
+
+			err = settings_register(&sh);
+			if (err) {
+				LOG_ERR("Cannot register settings handler");
+				goto error;
+			}
+		}
+
+		err = settings_load();
+		if (err) {
+			LOG_ERR("Cannot load settings");
+			goto error;
+		}
+		LOG_INF("Settings loaded");
+	}
+
+error:
+	return err;
+}
+
+static void init(void)
+{
+	int err = enable_settings();
+
+	if (err) {
+		module_set_state(MODULE_STATE_ERROR);
+	} else {
+		module_set_state(MODULE_STATE_READY);
+	}
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (IS_ENABLED(CONFIG_DESKTOP_CONFIG_CHANNEL_ENABLE)) {
+		if (is_config_event(eh)) {
+			const struct config_event *event = cast_config_event(eh);
+
+			if (!event->stored) {
+				int err = update_config(event->id, event->data);
+				if (err) {
+					LOG_ERR("Cannot store value, err %d", err);
+				}
+			}
+
+			return false;
+		}
+	}
+
+	if (is_module_state_event(eh)) {
+		const void * const req_modules[] = {
+#if CONFIG_BT_GATT_HIDS
+			MODULE_ID(hids),
+#endif
+#if CONFIG_DESKTOP_BAS_ENABLE
+			MODULE_ID(bas),
+#endif
+#if CONFIG_DESKTOP_BLE_ADVERTISING_ENABLE
+			MODULE_ID(ble_adv),
+#endif
+#if CONFIG_DESKTOP_MOTION_OPTICAL_ENABLE
+			MODULE_ID(motion),
+#endif
+		};
+		static unsigned int module_ready_cnt;
+		static bool initialized;
+
+		struct module_state_event *event = cast_module_state_event(eh);
+
+		/* Settings need to be loaded after GATT services are setup,
+		 * otherwise the values stored in flash will not be written
+		 * to GATT database.
+		 */
+		if (!initialized) {
+			for (size_t i = 0; i < ARRAY_SIZE(req_modules); i++) {
+				if (check_state(event, req_modules[i],
+				    MODULE_STATE_READY)) {
+					module_ready_cnt++;
```

Comment:
I know this is copy-paste but I was never fond of this method as it does not assert some module was not initiated twice.
Maybe we could try following:
```
static unsigned int req_state = 0;
for (size_t i = 0; i < ARRAY_SIZE(req); i++) {
   if (check_state(event, req[i], MODULE_STATE_READY)) {
      unsigned int flag = 1 << i;
      assert(req_state & flag == 0);
      req_state |= flag;

      if (req_state == BIT_MASK(ARRAY_SIZE(req))) {
         initme();
      }
      break;
   }
}
```

Reply:
I like this approach, done.